lib/elixir/lib/calendar/naive_datetime.ex

```diff
@@ -289,6 +289,27 @@ defmodule NaiveDateTime do
   end
 
   @doc """
+  Returns the given naive datetime with the microsecond field truncated to the
+  given precision (`:microsecond`, `millisecond` or `:second`).
+
+  ## Examples
+
+      iex> NaiveDateTime.truncate(~N[2017-11-06 00:23:51.123456], :microsecond)
+      ~N[2017-11-06 00:23:51.123456]
+
+      iex> NaiveDateTime.truncate(~N[2017-11-06 00:23:51.123456], :millisecond)
+      ~N[2017-11-06 00:23:51.123]
+
+      iex> NaiveDateTime.truncate(~N[2017-11-06 00:23:51.123456], :second)
+      ~N[2017-11-06 00:23:51]
+
+  """
+  @spec truncate(t, :microsecond | :millisecond | :second) :: t
```

Comment:
to match implementation and `truncate` on other modules, this should be:

```elixir
@spec truncate(Calendar.naive_datetime(), :microsecond | :millisecond | :second) :: Calendar.naive_datetime()
```

Btw, perhaps it actually could be:

```elixir
@spec truncate(Calendar.naive_datetime(), :microsecond | :millisecond | :second) :: t
```

This would match e.g.: https://github.com/rlustin/elixir/blob/v1.5.2/lib/elixir/lib/calendar/time.ex#L415

Reply:
Thanks for you feedback @wojtekmach. I’ve updated the commit with your first suggestion.

Reply:
just to illustrate why the 2nd suggestion might be better let's take a concrete example. Currently its:

```elixir
iex> Time.add(DateTime.utc_now(), 0)
~T[12:18:35.837717]
iex> Time.truncate(DateTime.utc_now(), :second)
#DateTime<2017-11-08 12:18:53Z>
```

I think it'd be more consistent to do:

```elixir
iex> Time.truncate(DateTime.utc_now(), :second)
~T[12:18:35]
```

Otherwise we're not gaining much by having separate truncate function on each struct (which have _identical_ implementations) because we already have `Time.truncate/2` that works on all structs. And if we really need a function that works on all structs then we could change `Calendar.truncate/2` to do just that.

I don't think it's a big deal in practice either way but just wondering. WDYT @josevalim?